File: homework_5/hw5.cpp
//
// Created by jugos000 on 09-Nov-20.
//

#include <cmath>
#include <iostream>
#include <iomanip>

#include "NelderMead.cpp"

using namespace std;

constexpr double Rc = 1000;
constexpr double Re = 1000;
//constexpr double E = 5;
constexpr double Vcc = 12;
constexpr double Is = 1e-13;
constexpr double Vt = 0.026;
constexpr double beta_f = 100;
constexpr double beta_r = 2;

double optimizationFunction(double E, double x[]) {
    const double Vbe = x[1];
    const double Vbc = x[2];

    double Ic = Is * ((exp(Vbe / Vt) - exp(Vbc / Vt)) - (exp(Vbc / Vt) - 1) / beta_r);
    double Ib = Is * ((exp(Vbe / Vt) - 1) / beta_f + (exp(Vbc / Vt) - 1) / beta_r);
    double Ie = Ic + Ib;

    double term1 = E - Vbe - Re * Ie;
    double term2 = Vcc - Rc * Ic + Vbc - E;

    return abs(term1) + abs(term2);
}

#ifndef TEST
int main(int argc, char **argv) {
    for (double E = 0; E <= 5; E += 0.1) {
        cout << "------- E = " << E << " V -------------------------------------" << endl;
        // memory allocation
        double **p = new double *[N + 2];
        for (int i = 0; i < N + 2; i++)
            p[i] = new double[N + 1];
        double *y = new double[N + 2];
        double ftol = (double) 1e-5;
        int nfunk;

        // initial simplex
        for (int i = 0; i < N + 2; i++) {
            for (int j = 0; j < N + 1; j++) {
                p[i][j] = 0.0;
                if (i > 1 && j == i - 1)
                    p[i][i - 1] = 1.0;
            }
            y[i] = f(p[i]);
        }

        // optimization
        NMsimplex(E, p, y, N, ftol, optimizationFunction, &nfunk);

        // output
        for (int i = 1; i < N + 1; i++)
            printf("%2.5f ", p[1][i]);
        printf("f(xmin)=%2.2f", y[1]);
        printf("\n");
        printf("%d\n", nfunk);

        double Vbe = p[1][1];
        double Vbc = p[1][2];
        double Vce = Vbe - Vbc;
        double Ic = Is * ((exp(Vbe / Vt) - exp(Vbc / Vt)) - (exp(Vbc / Vt) - 1) / beta_r);
        double Ib = Is * ((exp(Vbe / Vt) - 1) / beta_f + (exp(Vbc / Vt) - 1) / beta_r);
        double Ie = Ic + Ib;

        cout << "Vbe: " << setprecision(5) << Vbe << " V" << endl;
        cout << "Vbc: " << setprecision(5) << Vbc << " V" << endl;
        cout << "Vce: " << setprecision(5) << Vce << " V" << endl;
        cout << "Ic: " << setprecision(5) << Ic << " A" << endl;
        cout << "Ib: " << setprecision(5) << Ib << " A" << endl;
        cout << "Ie: " << setprecision(5) << Ie << " A" << endl;

        // clean-up
        for (int i = 0; i < N; i++)
            delete[] p[i];
        delete[] p;

        cout << "---------------------------------------------" << endl;
    }

    return 0;
}
#endif

File: homework_5/hw5.py
import math
import matplotlib.pyplot as plt
import numpy as np

from scipy.optimize import minimize

############################################

Rc = 1000
Re = 1000
Vcc = 12
Is = 1e-13
Vt = 0.026
beta_f = 100
beta_r = 2


############################################

def optimization_function(x):
    Vbe = x[1]
    Vbc = x[2]

    Ic = Is * ((math.exp(Vbe / Vt) - math.exp(Vbc / Vt)) - (math.exp(Vbc / Vt) - 1) / beta_r)
    Ib = Is * ((math.exp(Vbe / Vt) - 1) / beta_f + (math.exp(Vbc / Vt) - 1) / beta_r)
    Ie = Ic + Ib

    term1 = E - Vbe - Re * Ie
    term2 = Vcc - Rc * Ic + Vbc - E

    return abs(term1) + abs(term2)


############################################

def optimization(E):
    x = np.array([1, 1, 1])
    res = minimize(optimization_function, x, method='nelder-mead', options={'xatol': 1e-8, 'disp': True})
    print(res.x)

    Vbe = res.x[1]
    Vbc = res.x[2]
    Vce = Vbe - Vbc
    Ic = Is * ((math.exp(Vbe / Vt) - math.exp(Vbc / Vt)) - (math.exp(Vbc / Vt) - 1) / beta_r)
    Ib = Is * ((math.exp(Vbe / Vt) - 1) / beta_f + (math.exp(Vbc / Vt) - 1) / beta_r)
    Ie = Ic + Ib

    return np.array([Vbe, Vbc, Vce, Ic, Ib, Ie, res.fun])


############################################

E = 5
res = optimization(E)
print('Initial run (E = 5)')
print('---------------------------------------------------')
print('Vbe: ', "{:.6f}".format(res[0]), ' V')
print('Vbc: ', "{:.6f}".format(res[1]), ' V')
print('Vce: ', "{:.6f}".format(res[2]), ' V')
print('Ic:  ', "{:.6f}".format(res[3]), ' A')
print('Ib:  ', "{:.6f}".format(res[4]), ' A')
print('Ie:  ', "{:.6f}".format(res[5]), ' A')
print('f:   ', "{:.6f}".format(res[6]))
print('---------------------------------------------------')

print('---------------------------------------------------')
Vbe_array = np.array([])
Ic_array = np.array([])
Vout_array = np.array([])

for E in np.arange(0.0, 10.0, 0.1):
    res = optimization(E)
    print('Traversal run (E =', "{:.6f}".format(E), ')')
    print('---------------------------------------------------')
    print('Vbe: ', "{:.6f}".format(res[0]), ' V')
    print('Vbc: ', "{:.6f}".format(res[1]), ' V')
    print('Vce: ', "{:.6f}".format(res[2]), ' V')
    print('Ic:  ', "{:.6f}".format(res[3]), ' A')
    print('Ib:  ', "{:.6f}".format(res[4]), ' A')
    print('Ie:  ', "{:.6f}".format(res[5]), ' A')
    print('f:   ', "{:.6f}".format(res[6]))
    print('---------------------------------------------------')

    Vbe_array = np.append(Vbe_array, res[0])
    Ic_array = np.append(Ic_array, res[3])
    Vout_array = np.append(Vout_array, Re * res[5] + res[2])

plt.subplot(311)
plt.plot(np.arange(0.0, 10.0, 0.1), Vbe_array)
plt.grid()
plt.xlabel(r'$E [V]$')
plt.ylabel(r'$V_{be} [V]$')

plt.subplot(312)
plt.plot(np.arange(0.0, 10.0, 0.1), Ic_array)
plt.grid()
plt.xlabel(r'$E [V]$')
plt.ylabel(r'$I_{c} [V]$')

plt.subplot(313)
plt.plot(np.arange(0.0, 10.0, 0.1), Vout_array)
plt.grid()
plt.xlabel(r'$E [V]$')
plt.ylabel(r'$V_{out} [V]$')

plt.show()
